### DISCUSSION TITLE
How to create a hollow collision shape?

### DISCUSSION THREAD

**Discussion by LifeDrainingCoding** [AlmasB/FXGL#1381] [Q&A] [answered]
I'm triyng to create a single Entity with one Hitbox that should process collisions  only on the stroke(border) of the shape .
For an example i've tried to create a ring with outer circle(Should have collision  only with stroke(border)) hitbox and inner circle hitbox with sensor(inside this circle shouldn't be any collision events), but any dynamic entities seems like just squize out of the ring, because i think there's a collision inside circle even i added sensor  to not process any kind of collision events.  I have an idea to create 
a hollow collision shape with rectangles or triangles. But i'm asking for a simplier way to do that. 
```
 public class Ring {
    private double strokeWidth;
    private double radius;
    private Entity ringEntity,sensor;
    private HitBox inner,outer;
    private PhysicsComponent physics, sensorPhysics;
    public Ring(Point2D pos, double radius, double strokeWidth) {
        this.radius = radius;
        this.strokeWidth = strokeWidth;
        initPhysics();
        initHitboxes();
        Circle circle = new Circle(radius);
        circle.setFill(null);
        circle.setStroke(Color.TEAL);
        circle.setStrokeWidth(strokeWidth);
        ringEntity=  FXGL.entityBuilder().at(pos).with(physics).view(circle).with(new CollidableComponent(true)).bbox(outer)
                .buildAndAttach();
        sensor = FXGL.entityBuilder().at(pos).with(sensorPhysics).bbox(inner)
                .with(new CollidableComponent(false)).buildAndAttach();
    }
    private void initHitboxes(){
        outer = new HitBox("outer",BoundingShape.circle(radius));
        inner = new HitBox("inner",BoundingShape.circle(radius-strokeWidth));
        sensorPhysics.addSensor(inner, new SensorCollisionHandler() {
            @Override
            protected void onCollisionBegin(Entity other) {
                super.onCollisionBegin(other);
            }

            @Override
            protected void onCollision(Entity other) {
                super.onCollision(other);
            }

            @Override
            protected void onCollisionEnd(Entity other) {
                super.onCollisionEnd(other);
            }
        });
    }
    private void initPhysics(){

        sensorPhysics = new PhysicsComponent();
        sensorPhysics.setFixtureDef(new FixtureDef().sensor(true));
        sensorPhysics.setBodyType(BodyType.STATIC);
        physics =  new PhysicsComponent();
        physics.setBodyType(BodyType.STATIC);
        physics.setFixtureDef(new FixtureDef().friction(0.5f).density(1f).restitution(1f));
    }
} 
```

**Comment by AlmasB** [comment]
Hi, could you provide the use case for this and preferably a simple drawing sketch that shows how it is used / how it should work? Thanks.

**Reply by LifeDrainingCoding** [reply]
When a game developer wants to create a dynamic hollow collision shapes with configurable stroke width. Example: There's a game.  hollow circle it's a player  with movement. Player can move and earn filled circles  by collision with it. Collisioned filled circles transfers into    players hollow circle with enabled physics(so gained filled circles moves with players and have inertia and affect to players movement). 
By the way that entity with hollow collision  shape  should have collision only on the stroke of hollow shape(inside and outside).
![Here's a sketch for this idea](https://github.com/user-attachments/assets/60ac9736-a93a-48bc-829e-250be568ceae)

I want to create game using only FXGL polygons and circles.

**Reply by AlmasB** [reply, marked as answer]
I see, if my understanding is correct, when you use `outer = new HitBox("outer",BoundingShape.circle(radius));`, you end up with a filled rigid circle, but what you actually want is an [annulus](https://en.wikipedia.org/wiki/Annulus_(mathematics)) (a doughnut like shape with a hole inside). FXGL's rigid body dynamics internally uses box2d, which has edge and chain shapes, which may help you, [take a look at their documentation](https://box2d.org/documentation/md__d_1__git_hub_box2d_docs_collision.html).

In FXGL, you can call `BoundingShape.XXXX` to get edge / chain shapes, instead of circles. See if it works.
